### ISSUE TITLE
Hang (10+ days) "establishing new GRPC streaming session"

### ISSUE THREAD

**Issue by jtikalsky** [kopia/kopia#3073]
I have a kopia server instance on one PC that backs up to a kopia server repository on another PC.  Today I noticed that I had not had any snapshots created in the last 10 days.  I found a task pending with only one log line:

```
15:00:00.688 establishing new GRPC streaming session (purpose=Source Manager Uploader) 
```

Additionally, after I clicked "Cancel" in the Web UI, and waited several minutes, the task is still "Canceling".  Only when I restarted the kopia server process was it able to return to operating normally.

There should really be a timeout on this sort of thing so users do not end up going indefinitely without backups without realizing.

**Comment by jtikalsky** [comment]
I've encountered this again.  This time when I needed a recent backup, and none was available.

This is dangerous.  Users are deceived into thinking they have backups, and they do not.

**Comment by lupusA** [comment]
Hi @jtikalsky, 

was the server available? Usually this happens, if the server is not available and a snapshot is triggered. 

Cheers,

**Comment by jtikalsky** [comment]
Quite possibly not 10 days earlier, but it was available when I restarted the `kopia server` on the client.  I was able to snapshot no problem.

If a server being momentarily unavailable causes a client to never back up again, that's quite dangerous.  A timeout should be added in that case.

**Comment by rekonnected** [comment]
I am also encountering this. Has happened several times thus far. No warning, no notification, just a lockup.

![CleanShot 2023-10-22 at 19 13 53@2x](https://github.com/kopia/kopia/assets/31298266/17e99a41-6abc-473b-bc12-1f95e1844d53)

If it is the case that temporary server unavailability is causing this, then something needs to be added to enable a periodic reestablishment attempt.

**Comment by jtikalsky** [comment]
It has happened to me again today, actually.  I was unable to restore a file from yesterday because no backups have been created in 9 days.  "Stops making backups indefinitely in response to a temporary network error" is a devastating problem for an automatic backup program.  It's a shame because kopia is quite nice otherwise.

**Comment by kmbt** [comment]
Experiencing same behavior. Backing up into a repository server hung up for 25 days. 

Version v0.15.0 dd1ebb11cfdb7902a2b87106dfeea038858bafb5 kopia/htmlui 44ad26dabec430b9a1d1cad30aef2b984f55e515 built on Thu Oct 19 06:32:07 UTC 2023 fv-az478-358

In the log I see only one message:
09:45:00.001 establishing new GRPC streaming session (purpose=Source Manager Uploader)

**Comment by srkhost** [comment]
![image](https://github.com/kopia/kopia/assets/24195709/23e61568-9008-4e97-9cd6-dd5c241bd02f)

Same error, restarting kopia server and cancelling the task solves it.

![image](https://github.com/kopia/kopia/assets/24195709/fa01c68f-0067-42fb-99a4-0b75a6c211d3)

**Comment by JLyne** [comment]
I also encountered this problem today. All the same symptoms as above, snapshot stuck for multiple hours with `establishing new GRPC streaming session (purpose=Source Manager Uploader)` as the only log, restarting fixes it.

I have since been able to reproduce it again locally, with a server and client on the same machine. I'm going to try a third time with a debugger attached to see if I learn anything, but that seems to rule out network issues as an explanation.

**Comment by FurmigaHumana** [comment]
Issue persists on Version v0.18.2 c70f1a1c1164ee8676f85f9a1cea6de0a782a3ae kopia/htmlui e927e3a9bfdc87a52494490c2d381963feb7a1a0 built on Tue Nov 19 23:37:49 UTC 2024 fv-az847-965

**Comment by JLyne** [comment]
After reproducing it again, it seems the client is leaving grpc connections open forever after creating them? When the issue occurs a stack trace shows exactly 250 goroutines in `getOrEstablishInnerSession`. If I start another client at this point it will work, so it seems the client is the problem here.

As above this can be reproduced quickly by configuring a client to backup many snapshots to a repository server every minute. 10 snapshots should take 25 minutes.

Here are stack traces for the server and client.

[server goroutines.txt](https://github.com/user-attachments/files/18708970/server.goroutines.txt)
[client goroutines.txt](https://github.com/user-attachments/files/18708972/client.goroutines.txt)

@jkowalski, is there any additional information you would need to help identify the problem?

**Comment by CornFly2016** [comment]
Happens here on MacOS, to reproduce I mostly need to let the Mac sleep for a while, and then once it resumes it will be stuck like this trying to access the repository.
Restarting kopia-ui solves the issue, but this is a manual process

**Comment by scolton99** [comment]
I see this issue all the time across almost all my kopia-ui clients -- would definitely appreciate a fix.

**Comment by jtikalsky** [comment]
@scolton99
I have had to set a timer to kill and restart the kopia client daily just so it heals from this bug eventually rather than effectively disabling backups forever. Kopia really shouldn't hang forever if the server is temporarily unavailable.

**Comment by scolton99** [comment]
> [@scolton99](https://github.com/scolton99) I have had to set a timer to kill and restart the kopia client daily just so it heals from this bug eventually rather than effectively disabling backups forever. Kopia really shouldn't hang forever if the server is temporarily unavailable.

Yeah absolutely not ideal. My laptops turn on and off all the time and it's not so big a deal, but for my desktop I have to have a cron script that runs on the backup server to check how long it has been since the last snapshot to remind me to check things to see if the client is hanging there.

Out of curiosity, are you using any sort of reverse proxy to hit the repository server? I am using Apache and I have long wondered if that is part of the problem -- it was pretty difficult to get Apache's HTTP/2 reverse proxy settings to play nicely with Kopia. But aside from this, everything _seems_ to work great at the moment...

**Comment by jtikalsky** [comment]
@scolton99
> Out of curiosity, are you using any sort of reverse proxy to hit the repository server? I am using Apache and I have long wondered if that is part of the problem -- it was pretty difficult to get Apache's HTTP/2 reverse proxy settings to play nicely with Kopia. But aside from this, everything _seems_ to work great at the moment...

I am not.  My clients have direct access to the kopia server via a VPN.

**Comment by CornFly2016** [comment]
I don't use a proxy either, and I managed to reproduce this issue, and fix it with this PR: https://github.com/kopia/kopia/pull/5185
